rpc: prevent circular deps by extracting into separate include

brakmic opened the pull request:
### This PR is only a draft. 

## The description below will change over time.

----

This PR removes a potential **circular dependency** that was discovered in travis while building binaries for this PR https://github.com/bitcoin/bitcoin/pull/18827

The affected sources/includes in that PR are _httprpc_ and _rpc/server_

My first "solution" was a _forward declare_ of a single function in _rpc/server.cpp_. However, that merely defeats the linter and could not count as a proper solution.

After some discussion https://github.com/bitcoin/bitcoin/pull/18827#pullrequestreview-404116986 I tried to extract certain classes and function declarations into a new include file, _rpc/interfaces.h_

Now, the circular dependency in that PR is gone, but as this constitutes a much bigger change than the PR wanted to make, I think it should better get its own PR.

The following classes and functions got moved into a new source file pair *rpc/interfaces.cpp* and *rpc/interfaces.h*:

```cpp
class RPCTimerBase
class RPCTimerInterface
class CRPCCommand
class CRPCTable

void SetRPCWarmupStatus(const std::string& newStatus);
void SetRPCWarmupFinished();

bool RPCIsInWarmup(std::string *outStatus);
bool IsRPCRunning();

int64_t GetStartupTime();
int RPCSerializationFlags();

std::string JSONRPCExecBatch(const JSONRPCRequest& jreq, const UniValue& vReq);

void RPCSetTimerInterface(RPCTimerInterface *iface);
void RPCSetTimerInterfaceIfUnset(RPCTimerInterface *iface);
void RPCUnsetTimerInterface(RPCTimerInterface *iface);
void RPCRunLater(const std::string& name, std::function<void()> func, int64_t nSeconds);

bool IsDeprecatedRPCEnabled(const std::string& method);
```

In `src/Makefile.am` there is a new library defined: **LIBBITCOIN_SERVER_INTERFACES** that contains the implementation and is linked with `bitcoind` and `bitcoin-qt` (test and bench libraries are linked as well).

The changes reduce the size of `rpc/server.cpp` substantially. What remains there are pure server management functionalities:

```cpp
namespace RPCServer
{
    void OnStarted(std::function<void ()> slot);
    void OnStopped(std::function<void ()> slot);
}

void StartRPC();
void InterruptRPC();
void StopRPC();
```

Another advantage coming from these changes is that we got rid of forward declarations like `class CRPCTable`, `class CRPCCommand` etc.

----

### Possible future changes in class structure and function grouping

Although this PR isn't aiming for more complex engineering changes, like introducing new classes and other structures, there might be some advantage if we'd in future *group* the "extracted" functions into new classes. Just like we have `CRPCTable` that carries table-oriented functions, or `CRPCCommand`, we could introduce a new class called `CRPCTimer`:

```cpp
class CRPCTimer {
public:
    /** Set the factory function for timers */
    void RPCSetTimerInterface(RPCTimerInterface *iface);
    /** Set the factory function for timer, but only, if unset */
    void RPCSetTimerInterfaceIfUnset(RPCTimerInterface *iface);
   /** Unset factory function for timers */
   void RPCUnsetTimerInterface(RPCTimerInterface *iface);
};
```

And *server-status* functions could be moved into a new class like this:

```cpp
class CRPCServerStatus {
public:
    void SetRPCWarmupStatus(const std::string& newStatus);
    void SetRPCWarmupFinished();
    bool RPCIsInWarmup(std::string *outStatus);
    bool IsRPCRunning();
    int64_t GetStartupTime();
private:
   [...here come mutexes, guards, atomics etc. ]
};
```
Then we could give access to their instances the same way we do with `CRPCtable tableRPC` now.

```cpp
extern CRPCTimer timerRPC;
extern CRPCServerStatus statusRPC;
```

sipa commented:
This is still just defeating the linter, I'm afraid. It assumes that a .h file and a .cpp with the same name form a unit.

The question it's trying to answer is "can you use module X without having module Y or the other way around", where a module is seen as defined by F.cpp + F.h (if both exist).

brakmic commented:
> This is still just defeating the linter, I'm afraid. It assumes that a .h file and a .cpp with the same name form a unit.
> 
> The question it's trying to answer is "can you use module X without having module Y or the other way around", where a module is seen as defined by F.cpp + F.h (if both exist).

What if I would create rpc/interfaces.cpp that contains all the implementations?

sipa commented:
> What if I would create rpc/interfaces.cpp that contains all the implementations?

If that works, it's a solution indeed.

brakmic commented:
> > What if I would create rpc/interfaces.cpp that contains all the implementations?
> 
> If that works, it's a solution indeed.

I have now extracted `CRPCCommand`, `CRPCTable`, various `RPCTimer*`-structures and `Status/WarmUp` functions from `rpc/server.cpp` into the new `rpc/interfaces.cpp`. 

I'm deliberately using the plural form "interfaces" to indicate that these are interfaces that deal with RPC Server, but could be later split up into even smaller units, if needed. 

Also, the `shutdown.h/*.cpp` got removed to get rid of this linker error:

```shell
 AR       libbitcoin_wallet.a
  CXXLD    bitcoin-wallet
  CXXLD    bitcoind
Undefined symbols for architecture x86_64:
  "StartShutdown()", referenced from:
      stop(JSONRPCRequest const&) in libbitcoin_common.a(libbitcoin_common_a-interfaces.o)
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

I think that moving the shutdown functionality into `rpc/interfaces` is acceptable, because other server-control functions are located there as well, like `IsRPCRunning` and others.

While I was working on `rpc/interfaces` I also got a chance to remove all those forward class declarations `class CRPCTable`, which btw. provoked warnings like these:

```shell
Making all in src
  CXX      wallet/libbitcoin_wallet_a-rpcwallet.o
  AR       libbitcoin_wallet.a
  CXXLD    bitcoind
  CXXLD    bitcoin-wallet
ld: warning: duplicate symbol '_tableRPC' in:
    libbitcoin_util.a(libbitcoin_util_a-interfaces.o)
    libbitcoin_wallet.a(libbitcoin_wallet_a-rpcdump.o)
ld: warning: duplicate symbol '_tableRPC' in:
    libbitcoin_wallet.a(libbitcoin_wallet_a-wallet.o)
    libbitcoin_wallet.a(libbitcoin_wallet_a-rpcwallet.o)
ld: warning: duplicate symbol '_tableRPC' in:
    libbitcoin_wallet.a(libbitcoin_wallet_a-rpcwallet.o)
    libbitcoin_util.a(libbitcoin_util_a-interfaces.o)
ld: warning: duplicate symbol '_tableRPC' in:
    libbitcoin_zmq.a(libbitcoin_zmq_a-zmqrpc.o)
    libbitcoin_wallet.a(libbitcoin_wallet_a-wallet.o)
ld: warning: duplicate symbol '_tableRPC' in:
    libbitcoin_zmq.a(libbitcoin_zmq_a-zmqrpc.o)
    libbitcoin_wallet.a(libbitcoin_wallet_a-rpcdump.o)
ld: warning: duplicate symbol '_tableRPC' in:
    libbitcoin_zmq.a(libbitcoin_zmq_a-zmqrpc.o)
[...snip...]
```

`rpc/interfaces.cpp` is part of `libbitcoin_server_a_SOURCES` in `src/Makefile.am`

And just to make it even more complex for me, I also stumbled upon yet another circular dependency. :)

```shell
node/transaction -> validation -> rpc/interfaces -> rpc/util -> node/transaction
```
But after I replaced `node/transaction.h` with `util/error.h` in `rpc/util.h` it disappeared.

I hope that this proposed solution makes more sense. There are other circular dependencies still, but those exist independently from the one I was chasing in the last few days.

```shell
Circular dependency: chainparamsbase -> util/system -> chainparamsbase
Circular dependency: index/txindex -> validation -> index/txindex
Circular dependency: policy/fees -> txmempool -> policy/fees
Circular dependency: qt/addresstablemodel -> qt/walletmodel -> qt/addresstablemodel
Circular dependency: qt/bitcoingui -> qt/walletframe -> qt/bitcoingui
Circular dependency: qt/recentrequeststablemodel -> qt/walletmodel -> qt/recentrequeststablemodel
Circular dependency: qt/sendcoinsdialog -> qt/walletmodel -> qt/sendcoinsdialog
Circular dependency: qt/transactiontablemodel -> qt/walletmodel -> qt/transactiontablemodel
Circular dependency: txmempool -> validation -> txmempool
Circular dependency: wallet/fees -> wallet/wallet -> wallet/fees
Circular dependency: wallet/wallet -> wallet/walletdb -> wallet/wallet
Circular dependency: policy/fees -> txmempool -> validation -> policy/fees
```

sipa commented on src/rpc/interfaces.h:
```diff
@@ -0,0 +1,175 @@
+
+static const unsigned int DEFAULT_RPC_SERIALIZE_VERSION = 1;
+
+static std::atomic<bool> fRequestShutdown(false);
```
Making this static and in a .h file means that every compilation unit (cpp file) that includes it will get its own copy of the variable. If you only want a single copy, it must be declared in a .cpp file, and defined in the corresponding .h file.

I also think it's very strange that shutdown is now merged into an RPC-specific module. Does nothing outside of RPC need it?

brakmic replied:
> Making this static and in a .h file means that every compilation unit (cpp file) that includes it will get its own copy of the variable. If you only want a single copy, it must be declared in a .cpp file, and defined in the corresponding .h file.
> 

Thanks for the hint. Will correct it now.

> I also think it's very strange that shutdown is now merged into an RPC-specific module. Does nothing outside of RPC need it?

I never wanted to remove shutdown.h/cpp, but as I could never get pass the linker error I spoke above, the only way for me was to include shutdown's four methods into rpc/interface.

I haven't encountered any problems with shutdown removed. Functional RPC-tests are passing.  

But I'll gladly revert it, if there is a way to avoid the linker error.

brakmic replied:
> it must be declared in a .cpp file, and defined in the corresponding .h file.
> 

I must admit that I don't know how to do this. Especially with const variables and atomics.

sipa replied:
I'm being stupid sorry - it's a static variable that's local to shutdown.cpp. It can't and doesn't need to go in shutdown.h.

I think you shouldn't need to touch this at all. It's perfectly fine as-as in shutdown.cpp. It also has no dependencies, so touching it shouldn't be needed for the purpose of breaking circular dependencies.

The issue you're seeing while linking must have to do with the organization into libraries and possibly the order in which those are passed to the linker. That should be solvable by potentially moving things between libraries, but it's hard to say what the exact issue is without tinkering with it myself.

brakmic replied:
> I'm being stupid sorry - it's a static variable that's local to shutdown.cpp. It can't and doesn't need to go in shutdown.h.

Thanks for the clarification! I was already questioning life and existence. :)

> 
> I think you shouldn't need to touch this at all. It's perfectly fine as-as in shutdown.cpp. It also has no dependencies, so touching it shouldn't be needed for the purpose of breaking circular dependencies.
> 
> The issue you're seeing while linking must have to do with the organization into libraries and possibly the order in which those are passed to the linker. That should be solvable by potentially moving things between libraries, but it's hard to say what the exact issue is without tinkering with it myself.

Ok, then I will bring back shutdown.h/cpp and then try to find a way.

ryanofsky commented:
Have't looked closely at this but I'd discourage using "prevent circular deps" as the main justification for a change. If a change makes sense you should be able to say what advantages it has without reference to the linter, and I don't see other advantages cited in the PR description.

The circular dependencies linter often suggests joining files together which make more sense separately and splitting things up which make more sense together. It's also rare that it can't be satisfied by just dropping unnecessary includes and using forward-declared types appropriately.

sipa commented:
@ryanofsky Circular dependency here is not in the meaning of "can code be compiled without a cycle in #includes" but in the meaning of "can this module be used at a semantic leval without this other module". Forward declares do not solve that, they only hide the ability to detect it.

The sort of circular dependencies we're talking about here are a symptom of badly organized code, and improving it is a worthwhile goal in my opinion - but not at all costs. Sometimes it requires introducing costly abstractions like installing callbacks/signals, which just aren't worth it. Often it can be solved by just moving things appropriately so that module dependencies form a DAG.

ryanofsky commented:
> The sort of circular dependencies we're talking about here are a symptom of badly organized code, and improving it is a worthwhile goal in my opinion - but not at all costs.

I think we agree. I just think if the code is badly organized and the PR improves the organization, the PR description should actually say what was wrong with the organization before and how it's improved. I'm definitely willing to believe there's bad organization here, but I think it'd be good to take a few minutes and spell out what the actual problem is before spending too much time chasing linker and linter errors.

> Sometimes it requires introducing costly abstractions like installing callbacks/signals, which just aren't worth it. Often it can be solved by just moving things appropriately so that module dependencies form a DAG.

Definitely agree. I think if someone is introducing virtual classes and callbacks for no other reason than to satisfy the linter they are doing something wrong. I think there are cases like the fee estimation / mempool circular dependency where having modules call back and forth directly, but still be separate modules is a simple and good solution (https://github.com/bitcoin/bitcoin/pull/15638#discussion_r268718912)

brakmic commented:
>the PR description should actually say what was wrong with the organization before and how it's improved. 

Yes, the description will be updated as soon as this PR becomes a *real one*. Currently, it's only a draft and the only proper description is my first (longer) [answer](https://github.com/bitcoin/bitcoin/pull/18871#issuecomment-624237736) to @sipa regarding latest changes. As this is still a moving target and I am right now trying to overcome the linking error with `shutdown.cpp` any description in the header of this PR would sooner or later become obsolete. So, better not changing it every few minutes as this would only annoy people.

> I think if someone is introducing virtual classes and callbacks for no other reason than to satisfy the linter they are doing something wrong. 

My goal is to **change nothing**, that is: no *magic* via extra calls or structures to circumvent obstacles whatsoever. As @sipa said, it's about creating modules that could be used independently from each other. Either I will succeed in creating proper binary modules out of existing structures and functions, or I will have to give up.

So I hope it'll work.

brakmic commented:
@sipa 

`shutdown.h/.cpp` is back. I've removed its functions from `rpc/interface.h/.cpp` accordingly. 
RPC functional tests are passing. Unit tests as well. Qt code compiles too and its test pass also.
This change introduced no new circular dependency that'd affect `rpc/interface`. 

That is, they stayed the same as before:

```shell
Circular dependency: chainparamsbase -> util/system -> chainparamsbase
Circular dependency: index/txindex -> validation -> index/txindex
Circular dependency: policy/fees -> txmempool -> policy/fees
Circular dependency: qt/addresstablemodel -> qt/walletmodel -> qt/addresstablemodel
Circular dependency: qt/bitcoingui -> qt/walletframe -> qt/bitcoingui
Circular dependency: qt/recentrequeststablemodel -> qt/walletmodel -> qt/recentrequeststablemodel
Circular dependency: qt/sendcoinsdialog -> qt/walletmodel -> qt/sendcoinsdialog
Circular dependency: qt/transactiontablemodel -> qt/walletmodel -> qt/transactiontablemodel
Circular dependency: txmempool -> validation -> txmempool
Circular dependency: wallet/fees -> wallet/wallet -> wallet/fees
Circular dependency: wallet/wallet -> wallet/walletdb -> wallet/wallet
Circular dependency: policy/fees -> txmempool -> validation -> policy/fees
```

brakmic commented:
> > What if I would create rpc/interfaces.cpp that contains all the implementations?
> 
> If that works, it's a solution indeed.

Meanwhile, I have created a separate shared library that contains the "extracted" functions. There are no forward declarations anymore. I've also updated the PR description to reflect all the changes, and when you have time, it'd help me a lot, if you could check them.

Am I on the right track? Did I miss something important? Or is this whole thing simply not worth the effort?

brakmic committed:
rpc: prevent circular deps by extracting into separate include
